vendor/engines/split_accounts/app/services/split_accounts/order_detail_list_transformer.rb

```diff
@@ -10,13 +10,36 @@ def initialize(order_details = [])
 
     def perform(options = {})
       options ||= {} # in case it comes in as nil
-      order_details.each_with_object([]) do |order_detail, results|
-        # We will need to refactor the general_reports_controller_spec in
-        # order to remove the `try` methods below.
+
+      nested_map(order_details) do |order_detail|
         if order_detail.account.try(:splits).try(:present?)
-          results.concat SplitAccounts::OrderDetailSplitter.new(order_detail, split_time_data: options[:time_data]).split
+          SplitAccounts::OrderDetailSplitter.new(order_detail, split_time_data: options[:time_data]).split
         else
-          results << order_detail
+          order_detail
+        end
+      end
+    end
+
+    private
+
+    # Wraps a lazy enumerator with another enumerator so that if the block returns
```

Comment:
Can you clarify this comment a bit? I'm a little confused, and also it looks like the lazy enumerator is on this outside, not the inside, of the wrapping.

I'm assuming this is good for speed reasons--does this save a query when you're trying to do the splits above?

Reply:
It wraps one lazy enumerator with another lazy enumerator. `find_each` works as a lazy enumerator, meaning if you loop `each` (or another enumerable method--most `Enumerable` methods use `each` under the hood) on it, it won't hit the database until you need it to. So if you have `OrderDetail.find_each(batch_size: 1000).first(500)`, it'll only do the one query.

How it affects the splits above is this: before, even with `find_each`, the `each_with_object([])` is what is considered "terminal" meaning that it will render out the entire array doing as many queries as necessary and loading them all into memory before actually doing the loop. It's similar to `to_a`. You can't call `(1..Float::Infinity).to_a` because it'll go on forever, but you can call `(1..Float::Infinity).first(100)`.

What this basically does is each time you loop over the `each`, if it sees the next item is as an array (as in the case of split charges), it'll get the first thing out of the array on this loop, and the next thing on the next loop. If it's just a single item, it'll return that thing.

Here's what it does to the whole process:

Before:
* Create the query
* Load every order detail into memory (not sure if this was queried in batches or not, but it definitely loaded everything into memory)
* Loop over and create split order details into a new array
* Loop over the new array and write to CSV

Now: 
* Create the query
* For each 1000 order details
  * For each order detail
    * If non-split, write the row to CSV
    * If split, generate the multiple fake order-details, loop over them and write them to CSV

So really, the major advantage with this is memory: we only ever have 1000 (plus split fake-order-details) in un-freeable memory at once. Also, it doesn't affect this very much because we're sending it via email, but it is significantly faster to writing the first byte to CSV.

Reply:
Okay, that's very helpful. Can you clarify this comment that they're both lazy enums, and also point out briefly that the purpose of this is to save memory? It's helpful for me to have why's in comments, not just what's.